Readme.rst
shortmag.net
-------------

.. image:: https://travis-ci.org/XayOn/shortmagnet.svg?branch=master
    :target: https://travis-ci.org/XayOn/shortmagnet

.. image:: https://coveralls.io/repos/github/XayOn/shortmagnet/badge.svg?branch=master
 :target: https://coveralls.io/github/XayOn/shortmagnet?branch=master

.. image:: https://badge.fury.io/py/shortmagnet.svg
    :target: https://badge.fury.io/py/shortmagnet


Simple magnet-oriented URL Shortener.

Using ``redis`` to store more than 700000 magnet links.


Launching the server
--------------------

::

    shortmagnet.

    Magnet-oriented URL Shortener.

    Usage: shortmagnet [options]

    Options:
    --config=CONFIG  Config file
    --host=HOST      Host to listen on [default: 0.0.0.0]
    --port=PORT      Port to listen on [default: 8080]


Usage
-----

Shortmag.net does not have an actual user interface.

You can issue a ``POST`` request against main url (``/``) with form data magnet=magnet:?xt=foobar&baz...
to request a short link, to wich the server will directly answer with a text response containing
the short name to be used

Or you can issue a ``GET`` request against /<short_name> to get the actual redirect, direclty and **without
any intermediate pages or ads**.


Creating Short links
++++++++++++++++++++


* **URL**

  /

* **Method:**

  `POST`

*  **Form Params**

   **Required:**

   `magnet=[complete_magnet_url]`

* **Success Response:**

  * **Code:** 200 <br />
    **Content:** `short_name`

* **Error Response:**
  * **Code:** 503


Getting magnet links
++++++++++++++++++++

* **URL**

  /<short_name>

* **Method:**

  `GET`

 * **Success Response:**

  * **Code:** 302 <br />


Examples
++++++++


Sample usage with python-requests::

    import requests
    short = requests.post('http://shortmag.net/', data={'magnet': 'magnet:foo'}).text
    requests.get(short)



ChangeLog
CHANGES
=======

* Cleanup
* Added interpreters
* Probably travis working now
* Trying to figure out why it fails on travis
* Fixed view
* Fixed
* 100% test coverage
* Initial implementation with BDD
* Initial bdd test
* Doc skeleton
* Initial commit


tests/unit/test_main.py
"""Main tests."""


try:
    from unittest.mock import MagicMock, patch
except ImportError:
    from mock import MagicMock, patch


def test_create_app():
    """Test app creation and redis init."""
    import shortmagnet
    with patch('shortmagnet.FlaskRedis') as redis:
        with patch('shortmagnet.Flask') as flask:
            results = shortmagnet.create_app()
            assert len(results) == 2
            flask.assert_called_with(shortmagnet.__name__)
            redis.assert_called_with()
            redis().init_app.assert_called_with(flask())
            assert results[0] == flask()
            assert results[1] == redis()


def test_redis_config():
    """Test redis config extraction."""
    import os.path
    import sys

    config = os.path.join(os.path.dirname(__file__),
                          '../', 'fixtures', 'main.conf')

    import shortmagnet

    app, redis = MagicMock(), MagicMock()
    app.config = {}

    with patch('shortmagnet.create_app', side_effect=((app, redis,),)):
        sys.argv = [sys.argv[0], '--config', config]
        shortmagnet.main()
        assert app.config['REDIS_URL'] == "foobar"

    with patch('shortmagnet.create_app', side_effect=((app, redis,),)):
        sys.argv = [sys.argv[0]]
        shortmagnet.main()
        assert app.config['REDIS_URL'] == "redis://localhost:6379/0"


def test_main_view():
    """Test main view with fake redis."""
    import shortmagnet
    from mockredis import MockRedis

    redis = MockRedis()
    with patch('shortmagnet.request') as request:
        request.method = "POST"
        request.form = {'magnet': 'magnet:foobar'}
        result = shortmagnet.index("magnet:foobar", redis)
        assert result
        request.method = "GET"
        response = shortmagnet.index(result, redis)
        assert response.status_code == 302
        assert response.location == "magnet:foobar"


def test_main_view_assert_error():
    """Test main view with fake redis."""
    import pytest
    import shortmagnet
    from mockredis import MockRedis

    redis = MockRedis()
    with patch('shortmagnet.request') as request:
        with pytest.raises(AssertionError):
            request.method = "POST"
            request.form = {'magnet': 'foobar'}
            shortmagnet.index("foobar", redis)


tests/unit/test_package.py
"""Package related tests."""


def test_import():
    """Test import."""
    import shortmagnet  # noqa


tests/features/steps/steps_main.py
"""Main steps."""

from multiprocessing import Process
from behave import given, when, then  # noqa
import requests


@given(u'I have a shortener server up')
def setup_server(context):
    """Start server up."""
    from contextlib import closing
    from shortmagnet import main
    import socket

    def find_free_port():
        """Find a free port."""
        with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sck:
            sck.bind(('', 0))
            return sck.getsockname()[1]

    port = find_free_port()

    context.base_url = "http://localhost:{}/".format(port)
    import sys

    sys.argv = [sys.argv[0], '--port', str(port)]

    context.server = Process(target=main)
    context.server.start()
    import time
    time.sleep(2)


@when(u'I make a {request} request to {url}')
@when(u'I make a {request} request to {url} with {data}')
def make_request(context, request, url, data=None):
    """Make a request of type request to given url."""
    import json

    if not data:
        try:
            context.result = getattr(requests, request.lower())(
                context.base_url + context.result.text)
        except requests.exceptions.InvalidSchema as msg:
            context.exception = str(msg)
    else:
        context.result = getattr(requests, request.lower())(
            context.base_url, data=json.loads(data))


@then(u'I receive a body with a string')
def string_in_body(context):
    """Check body."""
    assert len(context.result.text) < 7, context.result.text


@then(u'I get a redirect to {place}')
def redirect_found(context, place):
    """Assert we found a redirect."""
    assert place in context.exception


@then(u'I stop the server')
def stop_server(context):
    """Stop server."""
    context.server.terminate()


shortmagnet/__init__.py
"""shortmag.net.

Magnet-oriented URL Shortener.
"""

import uuid
import configparser
from functools import partial
from docopt import docopt
from flask import Flask, request, redirect
from flask_redis import FlaskRedis


def create_app():
    """Create flask APP."""
    redis_store = FlaskRedis()
    app = Flask(__name__)
    redis_store.init_app(app)
    return app, redis_store


def index(param=None, redis_store=None):
    """Handle main route."""
    if request.method == "GET":
        return redirect(redis_store.get(param))
    assert request.form['magnet'].startswith('magnet:')
    name = str(uuid.uuid4())[:5]
    redis_store.set(name, request.form['magnet'])
    return name


def main():
    """shortmagnet.

    Magnet-oriented URL Shortener.

    Usage: shortmagnet [options]

    Options:
        --config=CONFIG  Config file
        --host=HOST      Host to listen on [default: 0.0.0.0]
        --port=PORT      Port to listen on [default: 8080]
    """
    options = docopt(main.__doc__)

    # Read config
    config = configparser.ConfigParser()
    config.add_section('main')
    config['main']['redis'] = "redis://localhost:6379/0"
    if options['--config']:
        config.read(options['--config'])

    # Create app and configure it
    app, redis_store = create_app()
    app.config['REDIS_URL'] = config['main']['redis']

    app.add_url_rule(
        '/<param>', 'index', partial(index, redis_store=redis_store),
        methods=["GET", "POST"])

    app.add_url_rule(
        '/', 'index_b', partial(index, redis_store=redis_store),
        methods=["GET", "POST"])

    # Run
    app.run(host=options['--host'], port=int(options['--port']))
